### PR TITLE
Add `array_equal`

### LINKED ISSUE #3125: implementing array_equal() for CuPy
I was trying to implement array_equal() for Cupy and was thinking along these lines

```py
def array_equal(a1, a2):
    try:
        a1, a2 = cupy.asarray(a1), cupy.asarray(a2)
    except Exception:
        return False
    if a1.shape != a2.shape:
        return False
    return bool(cupy.asarray(a1 == a2).all())
```

Should I move forward with this ? I would be grateful if someone could guide me.

### PR DESCRIPTION
**Pull request by rushabh-v**
closes #3125

### CODE REVIEW

File: cupy/logic/comparison.py [hunk 1]
```diff
@@ -38,6 +38,28 @@
 )
 
 
+def array_equal(a, b):
```
Comment (asi1024) [inline cupy/logic/comparison.py]: For compatibility to `numpy.array_equal`
```suggestion
def array_equal(a1, a2):
```

File: cupy/logic/comparison.py [hunk 2]
```diff
@@ -38,6 +38,28 @@
...
+    """Returns True if two arrays are element-wise exactly equal.
```
Comment (asi1024) [inline cupy/logic/comparison.py]:
```suggestion
    """Returns ``True`` if two arrays are element-wise exactly equal.
```

File: cupy/logic/comparison.py [hunk 3]
```diff
@@ -38,6 +38,28 @@
...
+
+    Returns:
+        cupy.ndarray: A boolean 0-dim array.
+            If its value is ``True``, two arrays are element-wise equal.
+
+    .. seealso:: :func:`numpy.array_equal`
+
+    """
+    a = cupy.asanyarray(a)
+    b = cupy.asanyarray(b)
```
Comment (asi1024) [inline cupy/logic/comparison.py]: The inputs are assumed to be `cupy.ndarray`, so this typecast is not needed.

File: cupy/logic/comparison.py [hunk 4]
```diff
@@ -38,6 +38,28 @@
...
+
+    if a.shape != b.shape:
+        return cupy.array(False)
+    return bool((a == b).all())
```
Comment (asi1024) [inline cupy/logic/comparison.py]: The return value should be 0-dim ndarray.
```suggestion
    return (a == b).all()
```

File: tests/cupy_tests/logic_tests/test_comparison.py [hunk 5]
```diff
@@ -74,6 +74,27 @@ def test_binary_array_pyscalar(self, xp, dtype):
...
+        a = xp.array([1, 2, 3, 4])
+        b = xp.array([1, 2, 3, 4])
+        return xp.allclose(a, b)
+
+    @testing.numpy_cupy_equal()
+    def test_array_equal_diff_length(self, xp):
+        a = xp.array([1, 2, 3, 4])
+        b = xp.array([1, 2, 3])
+        return xp.allclose(a, b)
+
```
Comment (asi1024) [inline tests/cupy_tests/logic_tests/test_comparison.py]: Please add tests check
- broadcasts are not allowed
- comparison with ndarrays of different dtypes is allowed

Comment (leofang) [reply tests/cupy_tests/logic_tests/test_comparison.py]: Why is allclose called instead of array_equal?

Comment (rushabh-v) [reply tests/cupy_tests/logic_tests/test_comparison.py]: Oh, I actually copy pasted the tests from `allclose` for my convenience and mistakenly left that call unchanged. I think I should've checked it..

I'll edit that soon.

File: cupy/logic/comparison.py (hunk 4 above)
Comment (rushabh-v) [reply cupy/logic/comparison.py]: Actually `(a == b).all()` returns `array(True)` or `array(False)`. And `numpy.array_equal` returns only `bool`s. So I thought they would not be comparable. 

Should I change that?

Comment (rushabh-v) [reply cupy/logic/comparison.py]: I'll make other changes soon :)

Comment (leofang) [reply cupy/logic/comparison.py]: Yes, you need to do this to follow CuPy's convention. This incompatibility is documented here: https://docs-cupy.chainer.org/en/stable/reference/difference.html#zero-dimensional-array

Comment (leofang) [reply cupy/logic/comparison.py]: (Note that `all()` *is* a kind of reduction operation.)

Comment (rushabh-v) [reply cupy/logic/comparison.py]: Thanks for providing with the documentation. 
I think I should read more and more developer guidelines before GSoC period for better understanding.

Comment (rushabh-v) [reply cupy/logic/comparison.py]: I was a bit busier today.
Will make all the changes soon.

**Comment by rushabh-v** [comment]
have made all the changes. And all the tests are passing.
Can you take a look, please? @asi1024

**Comment by asi1024** [comment]
@rushabh-v Could you check that broadcasts are not allowed?

**Comment by rushabh-v** [comment]
Added that test. Can you review it again, please? @asi1024

**Comment by rushabh-v** [comment]
I've just checked whether `Numpy` and `CuPy` results same or not for the broadcast case, because `numpy.array_equal` also doesn't allow broadcasts.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM. Thanks for your PR!

**Pull request merged**